firedrake/assemble.py

```diff
@@ -450,54 +460,32 @@ def apply_bcs(tensor, bcs, *, assembly_rank=None, form_compiler_parameters=None,
                 if V.component is None and V.index is not None:
                     # Mixed, index (no ComponentFunctionSpace)
                     if V.index == i:
-                        yield functools.partial(op2tensor[i, j].set_local_diagonal_entries, nodes)
+                        op2tensor[i, j].set_local_diagonal_entries(nodes)
                 elif V.component is not None:
                     # ComponentFunctionSpace, check parent index
                     if V.parent.index is not None:
                         # Mixed, index doesn't match
                         if V.parent.index != i:
                             continue
                         # Index matches
-                    yield functools.partial(op2tensor[i, j].set_local_diagonal_entries, nodes, idx=V.component)
+                    op2tensor[i, j].set_local_diagonal_entries(nodes, idx=V.component)
                 elif V.index is None:
-                    yield functools.partial(op2tensor[i, j].set_local_diagonal_entries, nodes)
+                    op2tensor[i, j].set_local_diagonal_entries(nodes)
                 else:
                     raise RuntimeError("Unhandled BC case")
-        for bc in equationbcs:
-            yield from _assemble(bc.f, tensor=tensor, bcs=bc.bcs,
-                                 form_compiler_parameters=form_compiler_parameters,
-                                 mat_type=mat_type,
-                                 sub_mat_type=sub_mat_type,
-                                 appctx=appctx,
-                                 assemble_now=assemble_now,
-                                 zero_tensor=False)
     elif assembly_rank == AssemblyRank.VECTOR:
-        for bc in dirichletbcs:
+        for bc in bcs:
             if assemble_now:
```

Comment:
Could someone please explain why we have this `assemble_now` parameter? Is the idea that boundary conditions are applied elsewhere in the solver so we can skip that here?

Reply:
It was introduced in d00719072 so I think it has some something to do with `EquationBC` objects, so I hope @ksagiyam can comment.

I don't really understand the logic. In fact, I think it's buggy. Consider:

```python
from firedrake import *
from firedrake.assemble import create_assembly_callable
mesh = UnitSquareMesh(1, 1, quadrilateral=True)
V = FunctionSpace(mesh, "Q", 1)
bc = DirichletBC(V, 1, "on_boundary")
v = TestFunction(V)
print(assemble(v*dx, bc=bc)).dat.data_ro)
# => [1, 1, 1, 1]
f = Function(V)
callable = create_assembly_callable(v*dx, tensor=f, bcs=[bc])
callable()
print(f.dat.data_ro)
# => [0, 0, 0, 0]
```

Reply:
We used to have `collect_loops` parameter, and `assemble_now == not collect_loops`. `assemble` returns a fully assembled matrix/vector if `assemble_now == True`, while it returns a callable if `assemble_now == False`.
Here, for vectors, I think we use `assemble_now` also a flag indicating what to assemble: if `True`, we assemble the solution, and, if `False`, we assemble the `residual`, which is why we get `[0,0,0,0]` in the Lawrence's example.

Reply:
> We used to have collect_loops parameter, and assemble_now == not collect_loops. assemble returns a fully assembled matrix/vector if assemble_now == True, while it returns a callable if assemble_now == False.

OK, this makes sense.

> Here, for vectors, I think we use assemble_now also a flag indicating what to assemble: if True, we assemble the solution, and, if False, we assemble the residual, which is why we get [0,0,0,0] in the Lawrence's example.

This is horrible and must be expunged.

Reply:
Yes, I agree.

Reply:
OK thanks. So what would be a nice way to do this?

Reply:
I think we should move to a case where `assemble_now` is not required.

So I think the low-level stuff should always be in `assemble_now == False` case which means that we're collecting the work to be done.

I don't understand enough about the BC application with equation bcs in the "residual evaluation" case. This happens inside `solving_utils._SNESContext.form_function`.

Reply:
When `_assemble` finds an `EquationBC` (in `apply_bcs`), whether it is assembling a residual or not, it first zeros associated rows of the vector/matrix, and then call `_assemble` from inside to fill those rows with the specified boundary equations. If the `EquationBC` itself has `DirichetBC`s (Dirichlet BCs of the Equation BC), those `DirichletBCs` are treated correctly in the second call to the `_assemble` according to if we are assembling a residual or a solution. BCs of BCs of `EquationBC`s and things like these are also treated in the same way with the recursive call of `_assemble`.

Yes, I think it's a good idea to remove `assemble_now` from `_assemble` (and related functions) with `assemble_now == False`. If `assemble_now == False`, `_assemble` currently automatically assembles "residual"s, so if we want to assemble a solution, we will need to do `bc.apply` somewhere at higher level, in which case we will need to do `bc.apply` recursively to set values of `DirichletBC`s of `DirichletBC`s ... of `EquationBC`s correctly.

Reply:
I've replaced `assemble_now` with a variable `assembly_type` which is either `"solution"` or `"residual"`. I may have misunderstood this conversation though but hopefully what I've done is valid.

Reply:
Can use an `IntEnum` for that, instead of a string?>